### PR TITLE
chore: convert Rooms Model Raw to TypeScript

### PR DESCRIPTION
**Pull request by totoi690**
Convert the file `Rooms.ts`, located on `/Rocket.Chat/apps/meteor/server/models/raw/Rooms.ts` to TypeScript

TC-495

### CODE REVIEW

File: apps/meteor/app/api/server/lib/rooms.ts [hunk 1]
```diff
@@ -181,7 +182,7 @@ export async function findRoomsAvailableForTeams({ uid, name }: { uid: string; n
 		Subscriptions.findByUserIdAndRoles(uid, ['owner'], { fields: { rid: 1 } }).fetch() as Pick<ISubscription, 'rid'>[]
 	).map((item) => item.rid);
 
-	const rooms = await Rooms.findChannelAndGroupListWithoutTeamsByNameStartingByOwner(uid, name, userRooms, options).toArray();
+	const rooms = await Rooms.findChannelAndGroupListWithoutTeamsByNameStartingByOwner(name, userRooms, options).toArray();
```
Comment (ggazzo) [inline apps/meteor/app/api/server/lib/rooms.ts:185]: why did you remove it? i'm quite sure uid is important here

File: packages/model-typings/src/models/IRoomsModel.ts [hunk 2]
```diff
@@ -1,107 +1,150 @@
...
-	findPaginatedRoomsWithoutDiscussionsByRoomIds(name: any, roomIds: any, options: any): any;
+	findPaginatedRoomsWithoutDiscussionsByRoomIds(
+		name: NonNullable<IRoom['name']>,
+		roomIds: Array<IRoom['_id']>,
+		options: FindOptions<IRoom>,
+	): FindPaginated<FindCursor<IRoom>>;
 
-	findChannelAndGroupListWithoutTeamsByNameStartingByOwner(uid: any, name: any, groupsToAccept: any, options: any): any;
+	findChannelAndGroupListWithoutTeamsByNameStartingByOwner(
+		name: NonNullable<IRoom['name']>,
```
Comment (ggazzo) [inline packages/model-typings/src/models/IRoomsModel.ts:74]: where is the uid ?

**Review by ggazzo — CHANGES_REQUESTED** [review]
if you haven't finished pr and it's not passing CI and not ready to be merged, consider opening your pr as Draft

File: apps/meteor/app/api/server/lib/rooms.ts (hunk 1 above)
Comment (totoi690) [reply apps/meteor/app/api/server/lib/rooms.ts:185]: I thought the uid was not being used inside the function, I'll double check

File: apps/meteor/app/livechat/server/hooks/offlineMessageToChannel.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ callbacks.add(
 			return data;
 		}
 
-		let channelName = settings.get('Livechat_OfflineMessageToChannel_channel_name');
+		let channelName = settings.get('Livechat_OfflineMessageToChannel_channel_name') as string;
```
Comment (matheusbsilva137) [inline apps/meteor/app/livechat/server/hooks/offlineMessageToChannel.ts:15]:
```suggestion
		let channelName = settings.get<string>('Livechat_OfflineMessageToChannel_channel_name');
```
We should avoid using `as Type`.

File: apps/meteor/app/livechat/server/hooks/offlineMessageToChannel.ts [hunk 4]
```diff
@@ -29,8 +29,8 @@ callbacks.add(
 			return data;
 		}
 
-		const room = await Rooms.findOneByName(channelName, { projection: { t: 1, archived: 1 } });
-		if (!room || room.archived || room.closedAt) {
+		const room: any = await Rooms.findOneByName(channelName, { projection: { t: 1, archived: 1 } });
+		if (!room || room.archived || room?.closedAt) {
```
Comment (matheusbsilva137) [inline apps/meteor/app/livechat/server/hooks/offlineMessageToChannel.ts:33]:
```suggestion
		const room = await Rooms.findOneByName(channelName, { projection: { t: 1, archived: 1 } });
		if (!room || room.archived || (isOmnichannelRoom(room) && room.closedAt)) {
```
The `closedAt` porperty is part of the `IOmnichannelGenericRoom` type.
Also add this import: `import { isOmnichannelRoom } from '@rocket.chat/core-typings';`

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]

File: apps/meteor/server/services/team/service.ts [hunk 5]
```diff
@@ -470,7 +470,7 @@ export class TeamService extends ServiceClassInternal implements ITeamService {
 			throw new Error('room-not-on-team');
 		}
 		room.teamDefault = isDefault;
-		Rooms.setTeamDefaultById(rid, isDefault);
+		void Rooms.setTeamDefaultById(rid, isDefault);
```
[the remark this answers is not in the document]
Comment (totoi690) [reply apps/meteor/server/services/team/service.ts:473]: In this case, as the user doesn't need to wait for the result, shouldn't we stick to using `void`?

File: apps/meteor/server/models/raw/Rooms.ts [hunk 6]
```diff
@@ -0,0 +1,1978 @@
...
+				$in: ids,
+			},
+			name: nameRegex,
+			t: type,
+		};
+
+		return this.find(query, options);
+	}
+
+	findGroupDMsByUids(uids: IRoom['uids'], options: FindOptions<IDirectMessageRoom> = {}): FindCursor<IDirectMessageRoom> {
```
Comment (matheusbsilva137) [inline apps/meteor/server/models/raw/Rooms.ts:1455]:
```suggestion
	findGroupDMsByUids(uids: NonNullable<IRoom['uids']>, options: FindOptions<IDirectMessageRoom> = {}): FindCursor<IDirectMessageRoom> {
```
Same here

File: apps/meteor/server/models/raw/Rooms.ts [hunk 7]
```diff
@@ -0,0 +1,1978 @@
...
+		const update: UpdateFilter<IRoom> = {
+			$set: {
+				lastMessage,
+			},
+		};
+
+		return this.updateOne(query, update);
+	}
+
+	async resetLastMessageById(_id: IRoom['_id'], lastMessage: IRoom['lastMessage'] | undefined): Promise<UpdateResult> {
```
Comment (matheusbsilva137) [inline apps/meteor/server/models/raw/Rooms.ts:1594]:
```suggestion
	async resetLastMessageById(_id: IRoom['_id'], lastMessage: IRoom['lastMessage']): Promise<UpdateResult> {
```
`lastMessage` can already be undefined

File: apps/meteor/server/models/raw/Rooms.ts [hunk 8]
```diff
@@ -0,0 +1,1978 @@
...
+			: {
+					$unset: {
+						lastMessage: 1,
+					},
+			  };
+
+		return this.updateOne(query, update);
+	}
+
+	replaceUsername(previousUsername: string, username: string): Promise<Document | UpdateResult> {
```
Comment (matheusbsilva137) [inline apps/meteor/server/models/raw/Rooms.ts:1612]:
```suggestion
	replaceUsername(previousUsername: IUser['username'], username: IUser['username']): Promise<Document | UpdateResult> {
```
That's essentially the same, but I guess it's better for maintanability

File: apps/meteor/server/models/raw/Rooms.ts [hunk 9]
```diff
@@ -0,0 +1,1978 @@
...
+		const update: UpdateFilter<IRoom> = {
+			$set: {
+				'usernames.$': username,
+			},
+		};
+
+		return this.updateMany(query, update);
+	}
+
+	replaceMutedUsername(previousUsername: string, username: string): Promise<Document | UpdateResult> {
```
Comment (matheusbsilva137) [inline apps/meteor/server/models/raw/Rooms.ts:1624]:
```suggestion
	replaceMutedUsername(previousUsername: IUser['username'], username: IUser['username']): Promise<Document | UpdateResult> {
```
Same here

File: apps/meteor/server/models/raw/Rooms.ts [hunk 10]
```diff
@@ -0,0 +1,1978 @@
...
+		const update: UpdateFilter<IRoom> = {
+			$set: {
+				'muted.$': username,
+			},
+		};
+
+		return this.updateMany(query, update);
+	}
+
+	replaceUsernameOfUserByUserId(userId: IRoom['u']['_id'], username: IRoom['u']['username']): Promise<Document | UpdateResult> {
```
Comment (matheusbsilva137) [inline apps/meteor/server/models/raw/Rooms.ts:1636]:
```suggestion
	replaceUsernameOfUserByUserId(userId: IUser['_id'], username: IUser['username']): Promise<Document | UpdateResult> {
```
It's the same, but we can use `IUser` directly

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]
Great PR! Congratsss

**Review by matheusbsilva137 — APPROVED** [review]

File: apps/meteor/server/services/team/service.ts [hunk 11]
```diff
@@ -364,7 +364,7 @@ export class TeamService extends ServiceClassInternal implements ITeamService {
 			room.teamId = teamId;
 		}
 
-		Rooms.setTeamByIds(rids, teamId);
+		void Rooms.setTeamByIds(rids, teamId);
```
Comment (KevLehman) [inline apps/meteor/server/services/team/service.ts:367]:
```suggestion
		await Rooms.setTeamByIds(rids, teamId);
```

File: apps/meteor/server/services/team/service.ts [hunk 12]
```diff
@@ -406,7 +406,7 @@ export class TeamService extends ServiceClassInternal implements ITeamService {
 
 		delete room.teamId;
 		delete room.teamDefault;
-		Rooms.unsetTeamById(room._id);
+		void Rooms.unsetTeamById(room._id);
```
Comment (KevLehman) [inline apps/meteor/server/services/team/service.ts:409]:
```suggestion
		await Rooms.unsetTeamById(room._id);
```

**Review by KevLehman — CHANGES_REQUESTED** [review]

**Review by matheusbsilva137 — APPROVED** [review]

File: apps/meteor/ee/server/lib/engagementDashboard/channels.ts [hunk 13]
```diff
@@ -40,6 +40,7 @@ export const findAllChannelsWithNumberOfMessages = async ({
 		end: convertDateToInt(end),
 		startOfLastWeek: convertDateToInt(startOfLastWeek),
 		endOfLastWeek: convertDateToInt(endOfLastWeek),
+		onlyCount: false,
```
Comment (KevLehman) [inline apps/meteor/ee/server/lib/engagementDashboard/channels.ts:43]: Maybe this should be optional? (Options should be optional too)

File: apps/meteor/app/slashcommands-inviteall/server/server.ts [hunk 14]
```diff
@@ -74,7 +74,7 @@ function inviteAll<T extends string>(type: T): SlashCommand<T>['callback'] {
 						lng,
 					}),
 				});
-			} else {
+			} else if (targetChannel) {
```
Comment (KevLehman) [inline apps/meteor/app/slashcommands-inviteall/server/server.ts:77]: Won't this have any change on the current logical flow?

File: apps/meteor/ee/server/lib/engagementDashboard/channels.ts (hunk 13 above)
Comment (matheusbsilva137) [reply apps/meteor/ee/server/lib/engagementDashboard/channels.ts]: I suggested @totoi690 to split it in two different methods: one for retrieving the results, and other one only for counting. Please check this later

File: apps/meteor/app/slashcommands-inviteall/server/server.ts (hunk 14 above)
Comment (matheusbsilva137) [reply apps/meteor/app/slashcommands-inviteall/server/server.ts]: Removed!

**Review by matheusbsilva137 — APPROVED** [review]
